docs/src/coding-guidelines.md

```diff
@@ -0,0 +1,130 @@
+2. Variable Names
+    - Should be nouns, as they represent state.
+    - Use *camel case* and start with an *lower-case* letter (e.g. overrideInputLayout)
+3. Function Names
+    - Represent actions and should be verb phrases
+    - Use *camel case* and start with a *lower-case* letter (e.g. createTTNNOptimizer(), emitTTNNAsCpp())
+
+# Includes
+
+We prefer #includes to be listed in this order:
+
+```
+1. Main Module Header
+2. Local/Private Headers
+3. LLVM project/subproject headers (clang/..., lldb/..., llvm/..., etc)
+4. System #includes
+```
+
+Using TTIRToTTNN.cpp as an example, this is what includes would look like for us:
+
+```
+#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"  # main header
+
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"  # these are local/private headers
+#include "ttmlir/Dialect/TTNN/Utils/Utils.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"  # llvm project/subproj headers
+#include "llvm/Support/LogicalResult.h"
+
+#include <cstdio>  # system includes
+#include <algorithm>
+```
+# Comments
+
+Write comments as full sentences, starting with a capital letter and ending with a period. Comments should explain why the code exists, not just what it does. Use comments to clarify logic, assumptions, or any non-obvious aspects of the code.
+
+Example of a comment:
+```c++
+// Initialize the buffer to store incoming data from the network.
+```
+
+In general use use C++-style comments (//). Use C-style comments (/**/) only for when documenting the significance of constants used as actual parameters in a call:
+```c++
+object.callFunction(/*arg0=*/nullptr);
+```
+
+Every function, class, or non-trivial piece of logic should have a comment. Avoid redundant comments for self-explanatory code, but never leave complex code unexplained. Example of redundant comment:
+```c++
+// Increment the counter by 1.  // Redundant, avoid.
+counter++;
+```
+
+Ensure comments are accurate and reflect the current state of the code. Outdated or misleading comments can be worse than no comments at all.
+
+# Code Denesting (Inversion)
+
+Strive to minimize unnecessary indentation without compromising code clarity. One effective way to achieve this is by using early exits and the continue keyword in long loops.
+
+Consider following example:
+```c++
+// ...
+if (modified && isDpsOp(op))
+{
+    // ... some long code ...
+}
+```
+
+It is strongly recommended to format the code as follows:
+```c++
+// ...
+// We just need to transform the op if it was modified
+if (!modified) return;
+
+// We also just need to transform the op if it is DPS op
+if (!isDpsOp(op)) return;
+
+// .. some long code ...
+```
+
+This reduces loop nesting, makes the reasoning behind the conditions clearer, and signals to the reader that there is no subsequent else to worry about, reducing cognitive load. This can significantly improve code readability and comprehension.
+
+# Using namespaces
+
```

Comment:
I don't know if this belongs to namespaces section, but every 'helper' function, i.e. function that's not part of the public interface and whose only purpose is to aid some other function in the same compilation unit should be either `static`, or wrapped inside an anonymous namespace. Besides obvious reason of potential linking error, it also serves as a good self-documenting feature of that function.

Reply:
If this kind of helper is in a cpp file I agree with `static`, but if this kind of helper is in a header it should be marked `inline`.